KAFKA-6342 : Move workaround for JSON parsing of non-escaped strings

umesh9794 opened the pull request:
This PR moves the JSON parsing workaround of [this PR](https://github.com/apache/kafka/pull/4303) to new method and uses this method in `ZkClient` etc. classes.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

umesh9794 committed:
Initial Commit

umesh9794 commented:
@ijuma Please review this PR. I have used the new method in `ZkUtils` however left `AdminUtils` as is. Please let me know your thoughts.

ijuma commented:
Thanks for the PR. I'm not sure I understand what was done here. We seem to be calling the new method from several places that are unrelated to ACLs? The idea was to limit the fallback to ACL handling code and the `Json` file would not have it any more.

umesh9794 commented:
@ijuma Thanks for the review. 
I tried to find out usage of current `kafka.utils.Json#parseFull` method but didn't find any ACL specific usage, however some `ZkUtils` methods were using this to parse ZK data e.g. topic data, leader and isr data. I thought to change `ZkUtils` to use the new `Json.parseFullIncludingACLs` method as it can return previously stored ACL JSON at some point. I was also suspected the "false assumption" for my this thinking. Sorry about the weird changes caused by this.

Next I tried to look at `ACL.scala` which uses `Json.parseBytes` however in the #4303 `Json.parseFull` method was changed. So am bit unsure at this point regarding the relevant usage of newly created ACL specific JSON parse method: `Json.parseFullIncludingACLs`. Probably I need your or @rajinisivaram's help to figure out next steps for this.

ijuma commented:
@umesh9794 Looks like the fix for #4303 only works for 1.0. As you say, in trunk  the code is using `parseBytes` instead. cc @rajinisivaram to confirm.

ijuma commented:
I suggest introducing a private method in `Acl` that parses the bytes and if there is a parsing error, it can then do a slow path where it converts the bytes to a string, replaces the potentially invalid characters and tries again.

umesh9794 commented:
@ijuma thanks for the suggestion, makes sense. I also need @rajinisivaram's confirmation on `Json.parseBytes` in `Acl` class based on our above observations.

umesh9794 commented:
@rajinisivaram just a gentle ping on this.

rajinisivaram commented:
@umesh9794 Sorry for the delay in responding to this. Yes, adding a new method in `Acl` to do the conversion as @ijuma suggested makes sense. Thanks.

umesh9794 commented:
@rajinisivaram thanks for the input. Sure, I'll proceed with the suggestion and update the PR soon.

umesh9794 committed:
Fixed the applicability of new method based on trunk

umesh9794 committed:
Merge branch 'trunk'

umesh9794 commented:
@rajinisivaram I have updated the PR based on the guidelines. However I've not changed `Acl` class much, rather I've changed the `Json` class to deal with the parsing error as suggested by @ijuma and implemented by you initially. Please review and let me know your comments.

rajinisivaram commented on core/src/main/scala/kafka/utils/Json.scala:
```diff
@@ -30,17 +32,26 @@ object Json {
-        val escapedInput = input.replaceAll("\\\\", "\\\\\\\\")
+        val escapedInput = new String(input, StandardCharsets.UTF_8).replaceAll("\\\\", "\\\\\\\\")
         try Option(mapper.readTree(escapedInput)).map(JsonValue(_))
         catch { case _: JsonProcessingException => None }
```
Could we use `parseBytes(escapedInput)` instead of the above two lines?

umesh9794 replied:
Good point. I'll update the PR soon.

ijuma commented on core/src/main/scala/kafka/utils/Json.scala:
```diff
@@ -30,17 +32,26 @@ object Json {
+   * Parse a JSON string into a JsonValue if possible. `None` is returned if `input` is not valid JSON. This method is currently used
+    * to read the already stored invalid ACLs JSON which was persisted using older versions of Kafka (prior to Kafka 1.1.0). KAFKA-6319
+   */
+  def parseBytesIncludingACLs(input: Array[Byte]): Option[JsonValue] =
```
Why do we have this here instead of `Acl`?

ijuma replied:
FYI, there's a `tryParseBytes` method that retains the exception.

umesh9794 replied:
I tried to retain the json parsing related methods in the `Json` class. Let me know if moving this parsing logic to `Acl` would be more legitimate.

Yep, `tryParseBytes` throws the `JsonProcessingException` instead of `None`. I'll update the PR to use `tryParseButes` instead.

ijuma replied:
This logic is only needed by Acl so it's better to keep it contained as a private method.

umesh9794 replied:
Gotcha, I'll do that

umesh9794 committed:
Rebased and applied review comments

umesh9794 commented:
Thanks for the review @ijuma @rajinisivaram. I have updated the PR with review comments. Please have a look and let me know your thoughts.

umesh9794 commented:
@rajinisivaram , @ijuma a gentle reminder on this!

rajinisivaram commented:
retest this please

rajinisivaram commented on core/src/test/scala/unit/kafka/utils/JsonTest.scala:
```diff
@@ -47,15 +47,6 @@ class JsonTest {
-    val map = Map("foo1" -> """bar1\,bar2""", "foo2" -> """\bar""")
-    val encoded = Json.legacyEncodeAsString(map)
-    val decoded = Json.parseFull(encoded)
-    assertEquals(Json.parseFull("""{"foo1":"bar1\\,bar2", "foo2":"\\bar"}"""), decoded)
```
This part of the test could still be here - since that should be encoding/decoding correctly?

umesh9794 replied:
Sure, will keep it there.

rajinisivaram commented on core/src/test/scala/unit/kafka/utils/JsonTest.scala:
```diff
@@ -47,15 +47,6 @@ class JsonTest {
-
-    // Test strings with non-escaped backslash and quotes. This is to verify that ACLs
-    // containing non-escaped chars persisted using 1.0 can be parsed.
-    assertEquals(decoded, Json.parseFull("""{"foo1":"bar1\,bar2", "foo2":"\bar"}"""))
```
We need an equivalent for this decoding for ACLs, but it has not been added?

umesh9794 replied:
I have added similar in `AclTest`, please have a look in the updated PR. However I am not able to hit the exception case in new `parseBytesWithAclFallback` method as the Json is getting parsed correctly in trunk. Please advise.

rajinisivaram reviewed (commented):
@umesh9794 Thanks for the updates. Looks good, left a couple of comments about testing.

ijuma commented on core/src/main/scala/kafka/security/auth/Acl.scala:
```diff
@@ -68,12 +75,28 @@ object Acl {
+        // Note that this does not handle all special characters (e.g. non-escaped double quotes are not supported)
+        val escapedInput = new String(input, StandardCharsets.UTF_8).replaceAll("\\\\", "\\\\\\\\")
+        Json.tryParseBytes(escapedInput.getBytes(StandardCharsets.UTF_8))
+    }
```
I think this can be written more simply as:

```scala
  private def parseBytesWithAclFallback(input: Array[Byte]): Option[JsonValue] = {
    Json.tryParseBytes(input) match {
      case Left(e) =>
        val escapedInput = new String(input, StandardCharsets.UTF_8).replaceAll("\\\\", "\\\\\\\\")
        Json.parseFull(escapedInput)
      case Right(v) => Some(v)
   }
```

I didn't compile the code above, but it should give you a reasonable idea of the intent.

umesh9794 replied:
Thanks for the tip, I'll incorporate this in the PR.

umesh9794 committed:
Rebased and applied review comments

umesh9794 commented:
Thanks @rajinisivaram @ijuma, please review the updated PR and let me know your comments.

rajinisivaram commented on core/src/test/scala/unit/kafka/security/auth/AclTest.scala:
```diff
@@ -22,25 +22,29 @@ import kafka.utils.Json
-    "{  \"host\": \"host1\",  \"permissionType\": \"Deny\",  \"operation\":   \"Read\" ,  \"principal\": \"User:bob\"}  ]}"
+    "{  \"host\": \"host1\",  \"permissionType\": \"Deny\",  \"operation\":   \"Read\" ,  \"principal\": \"User:bob\"},  " +
+    "{  \"host\": \"host1\",  \"permissionType\": \"Deny\",  \"operation\":   \"Read\" ,  \"principal\": \"User:\\\\bob\"}, " +
+    "{  \"host\": \"host1\",  \"permissionType\": \"Deny\",  \"operation\":   \"Read\" ,  \"principal\": \"User:bob1\\\\bob2\"}]}"
```
These two are still properly escaped strings. We should be testing `User:\\bob` and `"User:bob1\\bob2`? I think this Json would work even without the fallback code.

umesh9794 replied:
Sure, I tried with the below json: 

` "{  \"host\": \"host1\",  \"permissionType\": \"Deny\",  \"operation\":   \"Read\" ,  \"principal\": \"User:\\bob\"}, " + `
    
`"{  \"host\": \"host1\",  \"permissionType\": \"Deny\",  \"operation\":   \"Read\" ,  \"principal\": \"User:bob1\\bob2\"}]}" `

and following set : 

` val acl4 = new Acl(new KafkaPrincipal(KafkaPrincipal.USER_TYPE, "\\bob"), Deny, "host1", Read)`
 
`val acl5 = new Acl(new KafkaPrincipal(KafkaPrincipal.USER_TYPE, "bob1\\bob2"), Deny, "host1", Read) `

Still I don't see fallback code gets hit the execution and throws assertion error for below assert:

 `Assert.assertEquals(acls, Acl.fromBytes(AclJson.getBytes(UTF_8)))  `

`java.lang.AssertionError: 
Expected :Set(User:bob has Deny permission for operations: Read from hosts: host1, User:\bob has Deny permission for operations: Read from hosts: host1, User:bob1\bob2 has Deny permission for operations: Read from hosts: host1, User:bob has Allow permission for operations: Read from hosts: *, User:alice has Deny permission for operations: Read from hosts: host1)`

`Actual   :Set(User:bob1|bob2 has Deny permission for operations: Read from hosts: host1, User:bob has Deny permission for operations: Read from hosts: host1, User:|bob has Deny permission for operations: Read from hosts: host1, User:bob has Allow permission for operations: Read from hosts: *, User:alice has Deny permission for operations: Read from hosts: host1)`

I noticed that `Acl.fromBytes(jsonAcls)` has different values than `Acl.fromBytes(AclJson.getBytes(UTF_8))`. However completely not sure the possible reasons behind this.

umesh9794 replied:
@rajinisivaram, a gentle reminder on above message.

ijuma replied:
I recommend using `"""` to make it easier to understand what is being escaped.

umesh9794 replied:
@ijuma I tried using `"""`. Here is my test string: 
```
val AclJson = """{"version": 1, "acls": [{"host": "host1","permissionType": "Deny","operation": "READ", "principal": "User:alice"  },
    {  "host":  "*" ,  "permissionType": "Allow",  "operation":  "Read", "principal": "User:bob"  },
    {  "host": "host1",  "permissionType": "Deny",  "operation":   "Read" ,  "principal": "User:bob"},
    {  "host": "host1",  "permissionType": "Deny",  "operation":   "Read" ,  "principal": "User:\\bob"},
    {  "host": "host1",  "permissionType": "Deny",  "operation":   "Read" ,  "principal": "User:bob1\\bob2"}]}"""
```

And this string also doesn't hit the fall back code. It simply gets parsed without error and the escaped principals return back as below: 

```
User:\bob has Deny permission for operations: Read from hosts: host1
User:bob1\bob2 has Deny permission for operations: Read from hosts: host1 
```
Not completely sure about this behavior though. Any suggestions would be helpful.

umesh9794 committed:
Changed the AclJson to have the incorrect format for principals

ijuma commented on core/src/main/scala/kafka/utils/Json.scala:
```diff
@@ -30,20 +32,11 @@ object Json {
-   * Parse a JSON string into a JsonValue if possible. `None` is returned if `input` is not valid JSON.
-   */
+    * Parse a JSON string into a JsonValue if possible. `None` is returned if `input` is not valid JSON.
+    */
```
Let's not change this indentation as it makes it inconsistent with the rest of the file.

umesh9794 replied:
Ack.

ijuma commented:
@umesh9794, are you planning to address the last comments? I think we are ready to merge after that.

umesh9794 committed:
Applied the review comments

umesh9794 commented:
Sorry @ijuma I got distracted from this. I've updated the PR after applying the recent review comments. However as I notified earlier, the test is still not hitting the fallback code even though I am using suggested escaped sequence in ACL. Let me know if you see any obvious issues here. Thanks in advance.

viktorsomogyi commented:
@umesh9794 do you want to continue this PR or may I drive it to completion? We'd need this refactor in #4090.

umesh9794 commented:
@viktorsomogyi feel free to take over, thanks!

viktorsomogyi commented:
@ijuma can't force push to this branch so I couldn't rebase it. Created #8591 that continues the work that Umesh started. Would you please review that?
